crates/gpui/src/platform/linux/wayland/client.rs

```diff
@@ -63,8 +65,13 @@ pub(crate) struct WaylandClientStateInner {
     loop_handle: Rc<LoopHandle<'static, ()>>,
 }
 
+pub(crate) struct CursorStateInner {
+    cursor_icon_name: String,
+    cursor: Option<Cursor>,
+}
+
 #[derive(Clone)]
-pub(crate) struct WaylandClientState(Rc<RefCell<WaylandClientStateInner>>);
+pub(crate) struct WaylandClientState(Rc<RefCell<WaylandClientStateInner>>, Rc<RefCell<CursorStateInner>>);
```

Comment:
This cursor state should probably be added to the `WaylandClientStateInner`

Reply:
I wanted to track the cursor state as part of the wayland client state inner. But I couldn't borrow it in set_cursor_style(), because it complained "already borrowed". So I went with tracking the cursor state separately.

Reply:
This boils down to the state being mutable borrowed during the window updates (which end up calling `set_cursor_style` through some indirection), [here](https://github.com/zed-industries/zed/blob/main/crates/gpui/src/platform/linux/wayland/client.rs#L266) and [here](https://github.com/zed-industries/zed/blob/main/crates/gpui/src/platform/linux/wayland/client.rs#L288). Your solution seems reasonable to me, unless we manage to not hold that borrow during the `update` call (which seems possible but not trivial to me).

Reply:
I have the same issue in #8365, wayland event handlers borrow the state but then the actions caused by those events can call platform methods that also need the state. E.g. [key repeat with clipboard paste](https://github.com/zed-industries/zed/blob/9723ca95e3e959d44d347b449fdbbfd4690d273f/crates/gpui/src/platform/linux/wayland/client.rs#L504-L517) currently panics due to that issue, it would need to `drop(state)` before dispatching the event to stop the borrow (fixed in the clipboard PR).

I'm not sure how to solve but we'll need to refactor and split our state to avoid running into this issue.

But I don't like adding the cursor state at that location, `.1` doesn't sound semantic enough as a way to access the cursor state.

Reply:
> I have the same issue in https://github.com/zed-industries/zed/pull/8365, wayland event handlers borrow the state but then the actions caused by those events can call platform methods that also need the state. E.g. [key repeat with clipboard paste](https://github.com/zed-industries/zed/blob/9723ca95e3e959d44d347b449fdbbfd4690d273f/crates/gpui/src/platform/linux/wayland/client.rs#L504-L517) currently panics due to that issue, it would need to drop(state) before dispatching the event to stop the borrow (fixed in the clipboard PR).

Interesting.

> But I don't like adding the cursor state at that location, .1 doesn't sound semantic enough as a way to access it.

Ok, I have slightly reworked the WaylandClientState to have named variables and not as tuples like before. Does this look better? 

> I'm not sure how to solve but we'll need to refactor and split our state to avoid running into this issue. Maybe turning the state fields into rc-refcells.

I do agree that having all state fields as rc-refcells will save us some trouble, considering so many events are getting triggered in wayland and we wouldn't want to hold onto the entire state just because one operation is borrowing it longer. But, I am still new to Rust and all of this, so I don't know much about this. Will it be ok to do further refactoring in a separate MR or part of your Clipboard MR? If the approach isn't too bad/entirely wrong, I would like to get this in as an initial MR and further improvements can be worked upon it later.

Reply:
> I'm not sure how to solve but we'll need to refactor and split our state to avoid running into this issue. Maybe turning the state fields into rc-refcells.

Actually I'm not sure that would work. If I take as an example key press, I can imagine a scenario where we dispatch a key press on the `state.keyboard_focused_window`, which then triggers an action to change the keyboard focused window. So we'd get a panic there again. Not sure if there's a more elegant solution, but right now the only approach that works is to clone whatever refs we need before dropping the state, and then dispatching actions.

Or iiuc, the rest of the codebase seems to be using the pattern of passing a `cx` context down the stack, which avoids this kind of issue. We could also try that.